### PR TITLE
add support for collection variables

### PR DESCRIPTION
**Pull request by harryi3t**

### CODE REVIEW

**Comment by kunagpal** [comment]
Blocked on https://github.com/postmanlabs/postman-sandbox/pull/217

File: lib/runner/request-helpers-presend.js [hunk 1]
```diff
@@ -16,8 +16,12 @@ module.exports = [
 
         // @todo - resolve variables in a more graceful way
         // @todo - no need to sync vatiables when SDK starts supporting resolution from scope directly
-        context.item = new sdk.Item(context.item.toObjectResolved(null,
-            [context.data, context.environment.syncVariablesTo({}), context.globals.syncVariablesTo({})]));
+        context.item = new sdk.Item(context.item.toObjectResolved({variables: new sdk.VariableList(null, [])}, [
```
Comment (kamalaknn) [inline lib/runner/request-helpers-presend.js]: Can you add a `@todo` to remove the `variables` hack?

File: lib/runner/request-helpers-presend.js [hunk 2]
```diff
@@ -27,8 +31,13 @@ module.exports = [
 
         // resolve variables in auth
         auth && (context.auth = new sdk.RequestAuth(auth.toObjectResolved(
-            context.originalItem, // resolve using variables in item
-            [context.data, context.environment.syncVariablesTo({}), context.globals.syncVariablesTo({})]
+            {variables: new sdk.VariableList(null, [])}, // resolve using variables in item
```
Comment (kamalaknn) [inline lib/runner/request-helpers-presend.js]: Same as above ☝️

File: test/integration/sanity/collection-variable.test.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                    {key: 'echo-url', value: 'https://postman-echo.com', enabled: true}
+                ],
+                item: {
+                    name: 'Collection Variables Test Request',
+                    event: [{
+                        listen: 'test',
+                        script: {
+                            exec: [`
+                                console.log('test', {
+                                    'global-var': pm.variables.get('global-var'),
```
Comment (kamalaknn) [inline test/integration/sanity/collection-variable.test.js]: Could you update the tests to `pm.variables.toObject` once `postman-collection` is released?

File: test/integration/sanity/collection-variable.test.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                            type: 'basic',
+                            basic: [
+                                {key: 'username', value: '{{user}}'},
+                                {key: 'password', value: '{{pass}}'}
+                            ]
+                        }
+                    }
+                }
+            },
+            environment: {
```
Comment (kamalaknn) [inline test/integration/sanity/collection-variable.test.js]: Could you move this close to the `globals` definition?

File: lib/runner/replay-controller.js [hunk 5]
```diff
@@ -46,17 +46,12 @@ _.assign(ReplayController.prototype, /** @lends ReplayController.prototype */{
         !item.parent() && item.setParent(context.originalItem);
 
         // construct payload for request
-        var payload = _.defaults({
+        var payload = _.merge({}, desiredPayload, context, {
```
Comment (kamalaknn) [inline lib/runner/replay-controller.js]: @shamasis is it safe to do this or reconstruct context object? I think this allows `ReplayController` to have less knowledge of what `context` is made of.

File: lib/runner/extensions/waterfall.command.js [hunk 6]
```diff
@@ -51,11 +51,13 @@ module.exports = {
     init: function (done) {
         var state = this.state;
 
-        // ensure that the environment and globals are in VariableScope instance format
+        // ensure that the environment, globals and collectionVariables are in VariableScope instance format
         state.environment = VariableScope.isVariableScope(state.environment) ? state.environment :
             new VariableScope(state.environment);
         state.globals = VariableScope.isVariableScope(state.globals) ? state.globals :
             new VariableScope(state.globals);
+        state.collectionVariables = VariableScope.isVariableScope(state.collectionVariables) ?
```
Comment (kamalaknn) [inline lib/runner/extensions/waterfall.command.js]: Could you add a `@todo` to move this to item level, so we can support `Item` and `ItemGroup` variables?

Comment (kamalaknn) [reply lib/runner/extensions/waterfall.command.js]: Wrong place.

File: lib/runner/index.js [hunk 7]
```diff
@@ -49,6 +49,7 @@ _.assign(Runner.prototype, {
             data: Runner.normaliseIterationData(options.data, options.iterationCount),
             environment: options.environment,
             globals: _.has(options, 'globals') ? options.globals : this.options.globals,
+            collectionVariables: collection.variables,
```
Comment (kamalaknn) [inline lib/runner/index.js]: Could you add a @todo to move this to item level, so we can support Item and ItemGroup variables?

**Pull request merged**